File: src/main/js/ukrainiantolatin.js
/*
 * $Id$
 *
 * Copyright (c) 2012 Valentyn Kolesnikov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

Class.Mutators.Static = function(items){
    this.extend(items);
};
/**
 * UkrainianToLatin utility class.
 *
 * @author Valentyn Kolesnikov
 * @version $Revision$ $Date$
 */
var StringBuilder = new Class({
    initialize: function() {
        this._buffer = new Array();
    },

    append: function(text) {
            this._buffer[this._buffer.length] = text;
            return this;
    },
        
    insert: function(index, text) { 
            this._buffer.splice(index, 0, text);
            return this;
    },

    toString: function() {
            return this._buffer.join("");
    }
});

var ConvertCase = new Class({
        initialize: function(convert, lowcase) {
            this.convert = convert;
            this.lowcase = lowcase;
        }
});

var UkrainianToLatin = new Class({
    Static: {
        INDEX_0: 0,
        INDEX_1: 1,
        INDEX_2: 2,
        INDEX_3: 3,
        INDEX_4: 4,
        INDEX_8: 8,
        LENGTH_2: 2,
        LENGTH_3: 3,
        LENGTH_4: 4,
        LENGTH_8: 8,

        Convert: {
        AA:"Аа",
        BB:"Бб",
        VV:"Вв",
        HH:"Гг",
        GG:"Ґґ",
        DD:"Дд",
        EE:"Ее",
        YeIe:"Єє",
        ZhZh:"Жж",
        ZZ:"Зз",
        YY:"Ии",
        II:"Іі",
        YiI:"Її",
        YI:"Йй",
        KK:"Кк",
        LL:"Лл",
        MM:"Мм",
        NN:"Нн",
        OO:"Оо",
        PP:"Пп",
        RR:"Рр",
        SS:"Сс",
        TT:"Тт",
        UU:"Уу",
        FF:"Фф",
        KhKh:"Хх",
        TsTs:"Цц",
        ChCh:"Чч",
        ShSh:"Шш",
        ShchShch:"Щщ",
        YuIu:"Юю",
        YaIa:"Яя"
        },
    /**
     * Generates latinic from cyrilic.
     * @param name the name
     * @return the result
     */
    generateLat: function(name) {
        this.initialize();
        var result = new StringBuilder();
        var prevConvertCase = null;
        for (var index = 0; index < name.length; index += 1) {
            var curChar = name.substring(index, index + UkrainianToLatin.INDEX_1);
            var nextChar = index == name.length - 1 ? null : name.substring(index + UkrainianToLatin.INDEX_1, index + UkrainianToLatin.INDEX_2);
            if (curChar.match("[-'’,]")) {
                continue;
            }
            if (this.cyrToLat[curChar] == null) {
                if (" " == curChar) {
                    prevConvertCase = null;
                    result.append(' ');
                }
                continue;
            }
            var convertCase = this.cyrToLat[curChar];
            if (prevConvertCase == null) {
                this.checkFirstChar(result, convertCase, this.cyrToLat[nextChar] == null ? convertCase : this.cyrToLat[nextChar]);
            } else {
                this.checkMiddleChar(result, convertCase, this.cyrToLat[nextChar] == null ? convertCase : this.cyrToLat[nextChar]);
            }
            prevConvertCase = convertCase;
        }
        return result.toString();
    },
    checkFirstChar: function(result, convertCase, nextConvertCase) {
        var latName = convertCase.convert;
        switch (latName.length) {
        case UkrainianToLatin.LENGTH_2:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_1).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_1) : latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_1)
                    .toUpperCase());
            if (convertCase.convert == "ZZ" && nextConvertCase.convert == "HH") {
                result.append(nextConvertCase.lowcase ? "g" : "G");
            }
            break;
        case UkrainianToLatin.LENGTH_3:
        case UkrainianToLatin.LENGTH_4:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_2).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_2) : latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_2)
                    .toUpperCase());
            break;
        case UkrainianToLatin.LENGTH_8:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_4).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_4) : latName.substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_4)
                    .toUpperCase());
            break;
        default:
            break;
        }
    },
    /**
     * @param result
     * @param convertCase
     * @param prevChar
     */
    checkMiddleChar: function(result, convertCase, nextConvertCase) {
        var latName = convertCase.convert;
        switch (latName.length) {
        case UkrainianToLatin.LENGTH_2:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_1, UkrainianToLatin.INDEX_2).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_1, UkrainianToLatin.INDEX_2) : latName.substring(UkrainianToLatin.INDEX_1, UkrainianToLatin.INDEX_2)
                    .toUpperCase());
            if (convertCase.convert == "ZZ" && nextConvertCase.convert == "HH") {
                result.append(nextConvertCase.lowcase ? "g" : "G");
            }
            break;
        case UkrainianToLatin.LENGTH_3:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_3).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_3) : latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_3)
                    .toUpperCase());
            break;
        case UkrainianToLatin.LENGTH_4:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_4).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_4) : latName.substring(UkrainianToLatin.INDEX_2, UkrainianToLatin.INDEX_4)
                    .toUpperCase());
            break;
        case UkrainianToLatin.LENGTH_8:
            result.append(convertCase.lowcase ? latName.substring(UkrainianToLatin.INDEX_4, UkrainianToLatin.INDEX_8).toLowerCase() : nextConvertCase
                    .lowcase ? latName.substring(UkrainianToLatin.INDEX_4, UkrainianToLatin.INDEX_8) : latName.substring(UkrainianToLatin.INDEX_4, UkrainianToLatin.INDEX_8)
                    .toUpperCase());
            break;
        default:
            break;
        }
    },
    initialize: function() {
        this.cyrToLat = {};
        for (var convert in UkrainianToLatin.Convert) {
            this.cyrToLat[UkrainianToLatin.Convert[convert].substring(UkrainianToLatin.INDEX_0, UkrainianToLatin.INDEX_1)] = new ConvertCase(convert, false);
            this.cyrToLat[UkrainianToLatin.Convert[convert].substring(UkrainianToLatin.INDEX_1, UkrainianToLatin.INDEX_2)] = new ConvertCase(convert, true);
            if (convert == UkrainianToLatin.Convert.EE) {
                this.cyrToLat["Ё"] = new ConvertCase(convert, false);
                this.cyrToLat["ё"] = new ConvertCase(convert, true);
            }
        }
    }
    }
});

File: src/main/scala/UkrainianToLatin.scala
/*
 * $Id$
 *
 * Copyright 2015 Valentyn Kolesnikov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import collection.immutable.ListMap
import scala.collection.mutable.HashMap
import scala.util.control.Breaks._

/**
 * UkrainianToLatin utility class.
 *
 * @author Valentyn Kolesnikov
 * @version $Revision$ $Date$
 */
object UkrainianToLatin {
    val INDEX_0 = 0
    val INDEX_1 = 1
    val INDEX_2 = 2
    val INDEX_3 = 3
    val INDEX_4 = 4
    val INDEX_8 = 8
    val LENGTH_2 = 2
    val LENGTH_3 = 3
    val LENGTH_4 = 4
    val LENGTH_8 = 8

    val converts = ListMap(
        "AA" -> "Аа",
        "BB" -> "Бб",
        "VV" -> "Вв",
        "HH" -> "Гг",
        "GG" -> "Ґґ",
        "DD" -> "Дд",
        "EE" -> "Ее",
        "YeIe" -> "Єє",
        "ZhZh" -> "Жж",
        "ZZ" -> "Зз",
        "YY" -> "Ии",
        "II" -> "Іі",
        "YiI" -> "Її",
        "YI" -> "Йй",
        "KK" -> "Кк",
        "LL" -> "Лл",
        "MM" -> "Мм",
        "NN" -> "Нн",
        "OO" -> "Оо",
        "PP" -> "Пп",
        "RR" -> "Рр",
        "SS" -> "Сс",
        "TT" -> "Тт",
        "UU" -> "Уу",
        "FF" -> "Фф",
        "KhKh" -> "Хх",
        "TsTs" -> "Цц",
        "ChCh" -> "Чч",
        "ShSh" -> "Шш",
        "ShchShch" -> "Щщ",
        "YuIu" -> "Юю",
        "YaIa" -> "Яя")

    var cyrToLat : HashMap[String, ConvertCase] = null

    class ConvertCase(convert: String, lowcase: Boolean) {
        def getConvert() : String = {
            return convert
        }
        def isLowcase() : Boolean = {
            return lowcase
        }
    }

    cyrToLat = new HashMap[String, ConvertCase]()

    converts.foreach {case(key, value) =>
        cyrToLat(value.substring(INDEX_0, INDEX_1)) = new ConvertCase(key, false)
        cyrToLat(value.substring(INDEX_1, INDEX_2)) = new ConvertCase(key, true)
        if (key == "EE") {
            cyrToLat("Ё") = new ConvertCase(key, false)
            cyrToLat("ё") = new ConvertCase(key, true)
        }
    }

    /**
     * Generates latinic from cyrilic.
     * @param name the name
     * @return the result
     */
    def generateLat(name : String) : String = {
        if (name == null) {
            throw new IllegalArgumentException("name is null");
        }
        val result : StringBuilder = new StringBuilder()
        var prevConvertCase : ConvertCase = null
        for (index <- 0 until name.length()) {
            val curChar : String = name.substring(index, index + INDEX_1)
            val nextChar : String = if (index == name.length() - 1) null else name.substring(index + INDEX_1, index + INDEX_2)
            breakable {
                if (curChar.matches("[-'’,]")) {
                    break
                }
                if (cyrToLat.get(curChar).getOrElse(null) == null) {
                    if (" ".equals(curChar)) {
                        prevConvertCase = null
                        result.append(' ')
                    } else if (curChar.matches("\\n")) {
                        result.append(curChar)
                    }
                    break
                }
                val convertCase : ConvertCase = cyrToLat.get(curChar).getOrElse(null)
                if (prevConvertCase == null) {
                    checkFirstChar(result, convertCase, if (cyrToLat.get(nextChar).getOrElse(null) == null) convertCase else cyrToLat
                            .get(nextChar).getOrElse(null))
                } else {
                    checkMiddleChar(result, convertCase, if (cyrToLat.get(nextChar).getOrElse(null) == null) convertCase else cyrToLat
                            .get(nextChar).getOrElse(null))
                }
                prevConvertCase = convertCase
            }
        }
        return result.toString()
    }

    /**
     * Converts first character in the word.
     * @param result resut buffer to store string in latin
     * @param convertCase current character object
     * @param nextConvertCase next character object
     */
    def checkFirstChar(result : StringBuilder, convertCase : ConvertCase, nextConvertCase : ConvertCase) = {
        val latName : String = convertCase.getConvert()
        latName.length() match {
        case LENGTH_2 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_0, INDEX_1).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_0, INDEX_1) else latName.substring(INDEX_0, INDEX_1)
                    .toUpperCase())
            if (convertCase.getConvert() == "ZZ" && nextConvertCase.getConvert() == "HH") {
                result.append(if (nextConvertCase.isLowcase()) "g" else "G")
            }
        case LENGTH_3 | LENGTH_4 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_0, INDEX_2).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_0, INDEX_2) else latName.substring(INDEX_0, INDEX_2)
                    .toUpperCase())
        case LENGTH_8 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_0, INDEX_4).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_0, INDEX_4) else latName.substring(INDEX_0, INDEX_4)
                    .toUpperCase())
        }
    }

    /**
     * Converts middle or last character in the word.
     * @param result resut buffer to store string in latin
     * @param convertCase current character object
     * @param nextConvertCase next character object
     */
    def checkMiddleChar(result : StringBuilder, convertCase : ConvertCase, nextConvertCase : ConvertCase) = {
        val latName : String = convertCase.getConvert()
        latName.length() match {
        case LENGTH_2 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_1, INDEX_2).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_1, INDEX_2) else latName.substring(INDEX_1, INDEX_2)
                    .toUpperCase())
            if (convertCase.getConvert() == "ZZ" && nextConvertCase.getConvert() == "HH") {
                result.append(if (nextConvertCase.isLowcase()) "g" else "G")
            }
        case LENGTH_3 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_2, INDEX_3).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_2, INDEX_3) else latName.substring(INDEX_2, INDEX_3)
                    .toUpperCase());
        case LENGTH_4 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_2, INDEX_4).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_2, INDEX_4) else latName.substring(INDEX_2, INDEX_4)
                    .toUpperCase());
        case LENGTH_8 =>
            result.append(if (convertCase.isLowcase()) latName.substring(INDEX_4, INDEX_8).toLowerCase() else if (nextConvertCase
                    .isLowcase()) latName.substring(INDEX_4, INDEX_8) else latName.substring(INDEX_4, INDEX_8)
                    .toUpperCase());
        }
    }

    def main(args: Array[String]) {
        val message = "The utility class to convert ukrainian words to the latin characters.\n\n" +
        "For docs, license, tests, and downloads, see: https://github.com/javadev/ukrainiantolatin"
        print(message)
    }
}
